<a href="https://colab.research.google.com/github/karlmanalo/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS14_121_Statistics_Probability_Assignment_Manalo_Karl_200316.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

## 1. Load and clean the data (or determine the best method to drop observations when running tests)

In [0]:
### YOUR CODE STARTS HERE

import numpy as np
import pandas as pd
from scipy import stats

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

print(df.shape)
df.head()

--2020-03-16 19:41:14--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2020-03-16 19:41:14 (612 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df = df.replace({'y': 1, 'n': 0, '?': np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
rep = df[df['party'] == 'republican']
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [0]:
dem = df[df['party'] == 'democrat']
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

## 2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

In [0]:
stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

During the 1980s, Congressional Democrats tended to vote in favor of the budget more than Republicans.

## 3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

In [0]:
stats.ttest_ind(rep['physician-fee-freeze'], dem['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

During the 1980s, Congressional Republicans tended to vote in favor of freezing physician fees more than Democrats.

## 4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

In [0]:
stats.ttest_ind(rep['water-project'], dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

During the 1980s, there was not much difference between Congressional Republicans and Democrats towards water project cost sharing.

## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!

## 1. Refactor your code into functions so it's easy to rerun with arbitrary variables

In [0]:
# Not sure what exactly this means. Talk to Lori before turning in.

## 2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

In [0]:
# Assume we have a coin where the probability of landing on heads is .6. We will 
# flip this coin 10 times. The null hypothesis will be that this coin is fair
# and has a probability of landing on heads of .5.

import numpy as np
import pandas as pd
from scipy import stats

coinflips = np.random.binomial(1, p=.6, size=20)
ttest1_fakedata = pd.Series(coinflips)
ttest1_fakedata.describe()

count    20.00000
mean      0.65000
std       0.48936
min       0.00000
25%       0.00000
50%       1.00000
75%       1.00000
max       1.00000
dtype: float64

In [0]:
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=1.3708103001950636, pvalue=0.18641143545847966)

In [0]:
# Subtract sample mean from data.
ttest1_fakedata_sd1 = ttest1_fakedata.subtract(coinflips.mean())
ttest1_fakedata_sd1

0    -0.65
1     0.35
2    -0.65
3    -0.65
4    -0.65
5     0.35
6     0.35
7     0.35
8    -0.65
9     0.35
10    0.35
11    0.35
12   -0.65
13    0.35
14   -0.65
15    0.35
16    0.35
17    0.35
18    0.35
19    0.35
dtype: float64

In [0]:
# Square of previous step
ttest1_fakedata_sd2 = np.square(ttest1_fakedata_sd1)
ttest1_fakedata_sd2

0     0.4225
1     0.1225
2     0.4225
3     0.4225
4     0.4225
5     0.1225
6     0.1225
7     0.1225
8     0.4225
9     0.1225
10    0.1225
11    0.1225
12    0.4225
13    0.1225
14    0.4225
15    0.1225
16    0.1225
17    0.1225
18    0.1225
19    0.1225
dtype: float64

In [0]:
# Sum of previous step
sum(ttest1_fakedata_sd2)

4.549999999999998

In [0]:
# Divide by length of data
ttest1_fakedata_sd3 = sum(ttest1_fakedata_sd2)/len(coinflips)

In [0]:
# Square root of previous step
sd = np.sqrt(ttest1_fakedata_sd3)

sd

0.4769696007084727

In [0]:
ttest1_rewrite = (coinflips.mean() - .5) / ((sd)/(len(coinflips)**.5))

ttest1_rewrite

1.406421692815487

In [0]:
# Condensing above steps into one long line of code...
# np.sqrt(sum(np.square(ttest1_fakedata.subtract(coinflips.mean()))) / len(coinflips))
# The above line of code calculates the standard deviation of a given dataset.

ttest1_rewrite_final = (coinflips.mean() - .5) / (np.sqrt(sum(np.square(ttest1_fakedata.subtract(coinflips.mean()))) / len(coinflips)) / (len(coinflips)**.5))
ttest1_rewrite_final

1.406421692815487

I'm close to the right t-test value (I'll attribute any decimal level deviations to how Python is storing floats), but I'm not getting the right value for converting this value to a p-value.

Edit: Multiplying the p-value by 2 gives the correct value. Not sure what the reasoning is here though.

p-value based on function that I wrote:

.17574

p-value based on built-in scipy function:

.186

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

In [0]:
xbar1 = 180
s1 = 5

sample1 = pd.Series(np.random.normal(xbar1, s1, 1000))

In [0]:
xbar2 = 178.5
s2 = 4.25

sample2 = pd.Series(np.random.normal(xbar2, s2, 800))

In [0]:
stats.ttest_ind(sample1, sample2)

Ttest_indResult(statistic=6.454311789480462, pvalue=1.3944476960458865e-10)

In [0]:
# Cheating by copy-pasting my code from above for calculating standard 
# deviation. Still going to be a very long line of code.

ttest2_rewrite = (xbar1 - xbar2) / np.sqrt((np.sqrt(sum(np.square(sample1.subtract(sample1.mean()))) / len(sample1))**2) / 1000 + (np.sqrt(sum(np.square(sample2.subtract(sample2.mean()))) / len(sample2))**2) / 800)

ttest2_rewrite

6.927102068120821

Not sure if the difference between my method and Scipy's method is due to truncation of decimals or an error in the code.

Using this t-test value gives you a p-value of 0, expected with the given dataset where the means are definitely different given how many trials were performed.


 ### Then check your Answers using Scipy!

In [0]:
# Did this above within each respective section.